test/extensions/filters/udp/dns_filter/dns_filter_integration_test.cc

```diff
@@ -0,0 +1,170 @@
+      typed_config:
+        '@type': 'type.googleapis.com/envoy.extensions.filters.udp.dns_filter.v3alpha.DnsFilterConfig'
+        stat_prefix: "my_prefix"
+        server_config:
+          inline_dns_table:
+            external_retry_count: 3
+            known_suffixes:
+            - suffix: "foo1.com"
+            - suffix: "cluster_0"
+            virtual_domains:
+            - name: "www.foo1.com"
+              endpoint:
+                address_list:
+                  address:
+                  - 10.0.0.1
+                  - 10.0.0.2
+                  - 10.0.0.3
+                  - 10.0.0.4
+            - name: "cluster.foo1.com"
+              endpoint:
+                cluster_name: "cluster_0"
+    )EOF");
+  }
+
+  void setup(uint32_t upstream_count) {
+    udp_fake_upstream_ = true;
+    if (upstream_count > 1) {
+      setDeterministic();
+      setUpstreamCount(upstream_count);
+      config_helper_.addConfigModifier(
+          [upstream_count](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+            for (uint32_t i = 1; i < upstream_count; i++) {
+              bootstrap.mutable_static_resources()
+                  ->mutable_clusters(0)
+                  ->mutable_load_assignment()
+                  ->mutable_endpoints(0)
+                  ->add_lb_endpoints()
+                  ->mutable_endpoint()
+                  ->MergeFrom(ConfigHelper::buildEndpoint(
+                      Network::Test::getLoopbackAddressString(GetParam())));
+            }
+          });
+    }
+    BaseIntegrationTest::initialize();
+  }
+
+  void TearDown() override {
+    test_server_.reset();
+    fake_upstreams_.clear();
+  }
+
+  void requestResponseWithListenerAddress(const Network::Address::Instance& listener_address,
+                                          const std::string& data_to_send,
+                                          Network::UdpRecvData& response_datagram) {
+    Network::Test::UdpSyncPeer client(version_);
+    client.write(data_to_send, listener_address);
+    client.recv(response_datagram);
+  }
+
+  std::unique_ptr<DnsMessageParser> response_parser_;
+  DnsQueryContextPtr query_ctx_;
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, DnsFilterIntegrationTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+TEST_P(DnsFilterIntegrationTest, LocalLookupTest) {
+  setup(0);
+  const uint32_t port = lookupPort("listener_0");
+  const auto listener_address = Network::Utility::resolveUrl(
+      fmt::format("tcp://{}:{}", Network::Test::getLoopbackAddressUrlString(version_), port));
+
+  Network::UdpRecvData response;
+  std::string query =
+      Utils::buildQueryForDomain("www.foo1.com", DNS_RECORD_TYPE_A, DNS_RECORD_CLASS_IN);
+  requestResponseWithListenerAddress(*listener_address, query, response);
+
+  query_ctx_ = response_parser_->createQueryContext(response);
+  ASSERT_TRUE(query_ctx_->parse_status_);
```

Comment:
In general we want EXPECT for all of these. Only use ASSERT variants when the test absolutely cannot continue. Same elsewhere.

Reply:
Makes sense.  I will update.